### PR TITLE
Add decoding for fallback calls

### LINKED ISSUE #2212: Decode `fallback()` calls
# What is needed?
If a Safe transaction calls a `fallback()` function of a contract, this call is currently not decoded.

When we have a contract ABI available and no function matches the data, but we have an entry 
```json
{
  "type": "fallback",
  "stateMutability": "payable"
}
```
in the ABI, the decoding endpoint should return `fallback()` as function name without any parameters.

# Background
The` fallback()` function is a special function in Solidity that gets called if a contract is called with data that does not match any other existing function.

# Related issues
https://github.com/safe-global/safe-wallet-web/issues/4215

# Endpoint
I'd expect that the endpoint returns the `fallback()` function as called function:

```json
{
  "method": "fallback",
  "parameters": null
}
```

## URL
Example request:
```curl
curl -X 'POST' \
  'https://safe-transaction-mainnet.safe.global/api/v1/data-decoder/' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H 'X-CSRFToken: RzRiPGeNHIIyhyfmzMnbmwoW6QdRESx2lRGtr6ZeR76hzdkE2wTt5c2xXlYiflPA' \
  -d '{
  "data": "0x000001ad6abfb9ea000000000000000000000000000000000000000000000000000000000000019d0000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000098000000000000000000000000000000000000000000000000000000000000008e4ee8f0b86000000000000000000000000eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000016345785d8a000000000000000000000000000000000000000000000000000000000000000000cd00000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000808415565b0000000000000000000000000eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000016345785d8a0000000000000000000000000000000000000000000000000000000000000f1bd50a00000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000012000000000000000000000000000000000000000000000000000000000000004c000000000000000000000000000000000000000000000000000000000000005c0000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000040000000000000000000000000eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee000000000000000000000000000000000000000000000000016345785d8a000000000000000000000000000000000000000000000000000000000000000000210000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000034000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb4800000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000002c0000000000000000000000000000000000000000000000000016345785d8a0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000002556e6973776170563200000000000000000000000000000000000000000000000000000000000000016345785d8a0000000000000000000000000000000000000000000000000000000000000f21a484000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000f164fc0ec4e93095b804a4795bbe1e041497b92a00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001b000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000000000000000000000000000000000000005cf7a000000000000000000000000ad01c20d5886137e056775af56915de824c8fce5000000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee0000000000000000000000000000000000000000000000000000000000000000869584cd00000000000000000000000010000000000000000000000000000000000000110000000000000000000000000000000000000000ff8513c6b54542145a1b4cf70000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000b2c639c533813f4aa9d7837caf62653d097ff850000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000072a600000000000000000000000000000000000000000000000000000000000000cd00000000000000000000000000000000000000000000000000000000000000020000000000000000000000007e8485cf11c370519793d1c2d0a77bd139fdac38000000000000000000000000fea53c695fdf95cfb34514d916ac236e620201bd0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000f2ed992000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000066ed25970000000000000000000000000000000000000000000000000000000066ed799dd00dfeeddeadbeef765753be7f7a64d5509974b0d678e1e3149b02f4",
  "to": "0x3a23F943181408EAC424116Af7b7790c94Cb97a5"
}'
```

## Response
404 error (as function selector is not found)

### PR DESCRIPTION
**Pull request by falvaradorodriguez**
- Closes #2212

### CODE REVIEW

File: safe_transaction_service/contracts/tests/test_tx_decoder.py [hunk 1]
```diff
@@ -392,3 +392,33 @@ def test_db_tx_decoder(self):
         self.assertEqual(fn_name, "buyDroid")
         self.assertEqual(arguments, {"numberOfDroids": "4", "droidId": "10"})
         self.assertIn((contract.address,), DbTxDecoder.cache_abis_by_address)
+
+        #  Test fallback calls
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tests/test_tx_decoder.py:396]: Please do it in a different function. We should split this test also, I don't know why we put everything together

File: safe_transaction_service/contracts/tx_decoder.py [hunk 2]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
...
+    def contract_has_fallback_call(self, address: ChecksumAddress) -> bool:
+        """
+        :param address: Contract address
+        :return: `True` if the contract has a fallback call with `stateMutability` payable
+        """
+        abi = self.get_most_relevant_contract_abi(address)
+        if abi:
+            return any(
+                fn_abi.get("type") == "fallback"
+                and fn_abi.get("stateMutability") == "payable"
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:579]: Why must it be `payable`?

File: safe_transaction_service/contracts/tx_decoder.py [hunk 3]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
 
     cache_abis_by_address = TTLCache(maxsize=2048, ttl=60 * 5)  # 5 minutes of caching
 
+    @classmethod
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:555]: Why is this a `classmethod`?

File: safe_transaction_service/contracts/tx_decoder.py [hunk 4]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
...
+    def get_most_relevant_contract_abi(cls, address: ChecksumAddress):
+        """
+        Retrieves the most relevant ABI for the contract at the given address.
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:558]: This is working, but it's not accurate. `Contract address` is a `primary key`, so for one address there can be either `1` or `0` matching ABIs, so you don't need the sorting.

I don't know why we were doing it that way

File: safe_transaction_service/contracts/tx_decoder.py [hunk 5]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
 
     cache_abis_by_address = TTLCache(maxsize=2048, ttl=60 * 5)  # 5 minutes of caching
 
+    @classmethod
+    def get_most_relevant_contract_abi(cls, address: ChecksumAddress):
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:556]:
```suggestion
    def get_contract_abi(cls, address: ChecksumAddress) -> Optional[Dict[str, Any]]:
```

File: safe_transaction_service/contracts/tx_decoder.py [hunk 6]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
...
+        """
+        Retrieves the most relevant ABI for the contract at the given address.
+
+        :param address: Contract address
+        :return: List of ABI data if found, `None` otherwise.
+        """
+        return (
+            ContractAbi.objects.filter(contracts__address=address)
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:564]: In the case that this is required, this method should be a part of `ContractAbi` model `manager` and not of the decoder. But this is on me, as I implemented it that way before and you just moved the code

File: safe_transaction_service/contracts/tx_decoder.py [hunk 7]
```diff
@@ -552,6 +552,34 @@ class DbTxDecoder(TxDecoder):
...
+            .order_by("relevance")
+            .values_list("abi", flat=True)
+            .first()
+        )
+
+    def contract_has_fallback_call(self, address: ChecksumAddress) -> bool:
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:570]: This method is not using the ABI cache

File: safe_transaction_service/contracts/tx_decoder.py [hunk 8]
```diff
@@ -590,6 +614,12 @@ def get_abi_function(
                     # Otherwise we fall back to the general abi that matches the selector
                     return contract_selectors_with_abis[selector]
             return self.fn_selectors_with_abis[selector]
+        # Check if the contract has a fallback call and return a minimal ABIFunction for fallback call
+        elif address and self.contract_has_fallback_call(address):
+            return ABIFunction(
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:619]: You need to get the proper fallback function. Fallback can also take `bytes` as `input` or `output`: https://solidity-by-example.org/fallback/

**Review by Uxio0 — CHANGES_REQUESTED** [review]

File: safe_transaction_service/contracts/tx_decoder.py (hunk 2 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: You're right, my mistake! Fixed [here](https://github.com/safe-global/safe-transaction-service/pull/2262/commits/4caf91e5a242ebb07a6d3ff2cf11d225dbefc8da).

File: safe_transaction_service/contracts/tx_decoder.py (hunk 3 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: Completely unnecessary. My mistake.

File: safe_transaction_service/contracts/tx_decoder.py (hunk 4 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: Updated [here](https://github.com/safe-global/safe-transaction-service/pull/2262/commits/4caf91e5a242ebb07a6d3ff2cf11d225dbefc8da)

File: safe_transaction_service/contracts/tx_decoder.py (hunk 5 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: Updated [here](https://github.com/safe-global/safe-transaction-service/pull/2262/commits/4caf91e5a242ebb07a6d3ff2cf11d225dbefc8da)

File: safe_transaction_service/contracts/tx_decoder.py (hunk 6 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py:567]: Noted, for the moment I keep it here as it has no logic that requires it to be in the model manager.

File: safe_transaction_service/contracts/tx_decoder.py (hunk 7 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: I added a new chache for the abis by address and kept the cache with the selectors for the abi functions by address.

File: safe_transaction_service/contracts/tx_decoder.py (hunk 8 above)
Comment (falvaradorodriguez) [reply safe_transaction_service/contracts/tx_decoder.py]: Understood, Updated [here](https://github.com/safe-global/safe-transaction-service/pull/2262/commits/4caf91e5a242ebb07a6d3ff2cf11d225dbefc8da).

File: safe_transaction_service/contracts/tx_decoder.py [hunk 9]
```diff
@@ -551,22 +551,57 @@ class DbTxDecoder(TxDecoder):
...
+    def get_contract_fallback_function(
+        self, address: ChecksumAddress
+    ) -> Optional[ABIFunction]:
+        """
+        :param address: Contract address
+        :return: Fallback ABIFunction if found, `None` otherwise.
+        """
+        abi = self.get_contract_abi(address)
+        if abi:
+            return next(
```
Comment (Uxio0) [inline safe_transaction_service/contracts/tx_decoder.py:581]: Really nice, using functional programming

**Review by Uxio0 — APPROVED** [review]

**Pull request merged**